Issue2627 prepwork decouple selection expansion emancipation checklist

vasconsaurus opened the pull request:
### What github issue is this PR for, if any?
Progress on #2627

### What changed, and why?
Refactored the function to be more modular, so it's easier to track the collapse icon state and decouple actions. Right now it still works the same as before.

### How is this tested? (please write tests!) 💖💪
JS tests:

- will add some 🙂 

Rspec:

- spec/system/emancipations
- spec/views/emancipations
- spec/controllers/emancipations_controller_spec.rb
- spec/helpers/emancipations_helper_spec.rb

### Screenshots please :)
<img width="999" alt="Screen Shot 2022-10-18 at 23 24 23" src="https://user-images.githubusercontent.com/87862340/196582553-c8c6ea81-7319-43d7-87b0-64ff74dcdb0b.png">

vasconsaurus commented on app/javascript/__tests__/case_emancipations.test.js:
```diff
@@ -60,7 +60,7 @@ describe('Function that closes the children of a given parent', () => {
     $(document).ready(() => {
       closeChildren(category)
-      expect(category.attr('data-is-open')).toContain('false')
+      expect(category.attr('data-is-open')).toContain('banana')
```
@compwron 
some tests that should have failed passed when I ran `npm test` locally, so I decided to check if the failure would show up when I pushed, but it doesn't seem to show up anywhere as a failure.

How should I go about this?

compwron replied:
hmmmm that is concerning, but they are failing in CI... maybe your local is having caching issues
or maybe we are not running all the tests I should be

vasconsaurus replied:
so, I don't think `toContain` was the problem, as suggested [here](https://github.com/rubyforgood/casa/pull/4081/files/d0fe2fbc2df16fc0e4c35f7fe4db342a07ac5a04#r1002404117).

So, when I started to write the tests and ran `npm test`, they were failing when they should, but at some point, they stopped failing. Even if I comment my variables, and test for things like 'banana' it still passes.

Now it seems my case_emacipation.js stopped working (other js files work fine), even when I'm on the main branch, so I don't think any updates to the file on my branch are the problem. I think there is probably something wrong with my environment, but I'm at a loss about what it could be.

vasconsaurus replied:
@tcione helped me and we figured out what was wrong 😄 There are two different problems:

1- I followed the pattern from one of the test files that used `$(document).ready(() => {` inside the tests, turns out those tests aren't running. So the problem with my tests was they weren't running. There are other tests that aren't running as well. I'll open an issue explaining this better.

2- On why the js stopped working: there is a regex in `line 104` that tested only for digits in the pathname, but 
this is the pathname we get: '/casa_cases/CINA-07-1003/emancipation'. If I'm not mistaken, you talked about how the pathname was going to change last office hours. So that is why this also broke on main.

vasconsaurus replied:
I added two small issues for these bugs, and went ahead and fixed the regex one:
[issue#4107](https://github.com/rubyforgood/casa/issues/4107) and [issue#4108](https://github.com/rubyforgood/casa/issues/4108)/[PR for issue#4108](https://github.com/rubyforgood/casa/pull/4109)

compwron commented:
Yay, I am glad you are working on this! :)

tcione commented on app/javascript/__tests__/case_emancipations.test.js:
```diff
@@ -0,0 +1,75 @@
+  test('Deselects the inputs in the categoryOptionsContainer', () => {
+    $(document).ready(() => {
+      deselectChildren(category)
+      // expect(how do I get the input and check if it's not checked ?)
```
@vasconsaurus I think you could use [`.is()`](https://api.jquery.com/is/). Something like: `$element.is(':checked')`. You could also get the native DOM element and use vanilla js (something like `$element[0].checked`)

tcione commented on app/javascript/src/case_emancipation.js:
```diff
@@ -62,6 +62,43 @@ function saveCheckState (action, checkItemId) {
+}
+
+export function closeChildren ($parent) {
+  const categoryOptionsContainer = $parent.siblings('.category-options')
```
**nitpick**: the functions point that they receive a jquery element by using `$` at the start of the variable name (`$parents`). Then we store a different jquery element without using `$` (`categoryOptionsContainer`). I think being consistent would make code clearer (so either using `$` for all jquery elements (`$parents` and `$categoryOptionsContainer`, or using nothing)

tcione commented on app/javascript/src/case_emancipation.js:
```diff
@@ -62,6 +62,43 @@ function saveCheckState (action, checkItemId) {
     })
 }
 
+export function manageTogglerText ($parent) {
```
**suggestion:** since these methods usually should act under the same parent, it might be worthwhile to wrap this in a class or using closures.

Something like
```js
class Toggler {
  constructor($parent) {
    this.$parent = $parent
  }

  manageTogglerText() {
    // your method using this.$parent
  }
}

const $category = $('selector')
const toggler = new Toggler($category)
toggler.manageTogglerText()
```

or

```js
function Toggler($parent) {
  const manageTogglerText = () => {
    // your method using $parent
  }

  return { manageTogglerText }
}

const $category = $('selector')
const toggler = Toggler($category)
toggler.manageTogglerText()
```

vasconsaurus replied:
I had forgotten about this, but updated it to use a class

compwron commented on app/javascript/src/case_emancipation.js:
```diff
@@ -62,6 +62,43 @@ function saveCheckState (action, checkItemId) {
+export function manageTogglerText ($parent) {
+  const categoryCollapseIcon = $parent.find('.category-collapse-icon')
+
+  if ($parent.attr('data-is-open') === 'true') {
```
I don't understand how that would work but if it would work I am onboard!

tcione replied:
In the end it's bit silly and probably boils down to preference. But it would become something like `parent.data('is-open')`

vasconsaurus commented on app/javascript/__tests__/case_emancipations.test.js:
```diff
@@ -0,0 +1,75 @@
+      category.attr('data-is-open', 'false')
+
+      manageTogglerText(category)
+      expect(categoryCollapseIcon.text).toContain('+')
```
uuuuuu I'll take a look at this (an all the other suggestions 😉 ) thanks ❤️

vasconsaurus commented:
@compwron I'm finally done with this one 🙂 But while I was trying to test git rebase, I think I made a big mess of the commit history... I'm very sorry for the mess 😞

vasconsaurus committed:
made structure change and tests

I had to re-create these files because of a git mess I made.
So I lost the commits.

vasconsaurus commented:
> @compwron I'm finally done with this one 🙂 But while I was trying to test git rebase, I think I made a big mess of the commit history... I'm very sorry for the mess 😞

The problem was worse than just the commits, but it is fixed 🙃 
BUT I just found a small bug, so there is still that to fix 🫠

vasconsaurus commented:
> The problem was worse than just the commits, but it is fixed 🙃 BUT I just found a small bug, so there is still that to fix 🫠

@compwron It's DONE! Can you please review this for me? 
What I thought was a bug was actually the original behavior 🙃

vasconsaurus committed:
wrapped manageTogglerText, closeChildren and openChildren in a class

since these methods usually should act under the same parent,
it might be worthwhile to wrap this in a class or using closures.

vasconsaurus committed:
finished wrapping up the functions inside the class

also updated the tests

tcione commented on app/javascript/__tests__/case_emancipations.test.js:
```diff
@@ -0,0 +1,68 @@
+describe('Function that deselects the children of a deselected parent', () => {
+  test('Deselects the inputs in the categoryOptionsContainer', () => {
+    toggler.deselectChildren(() => '')
+    expect(checkBox.prop('checked')).toBe(false)
```
**nitpick:** `toBe` -> `toEqual` (for consistency)

tcione reviewed (approved):
![LGTM](https://image.lgtmoon.dev/177360)

Thanks for going through all the comments and changes. That said, my approval is merely symbolic, so please wait for @compwron review :D

vasconsaurus committed:
toBe -> toEqual (for consistency)
